### PR TITLE
Clear sender input witness before processing PSBT

### PR DESCRIPTION
**Pull request by grizznaut**
Based on ongoing conversation: https://github.com/payjoin/rust-payjoin/issues/51#issuecomment-2089150739

### CODE REVIEW

File: payjoin/src/receive/mod.rs [hunk 1]
```diff
@@ -559,27 +577,19 @@ impl ProvisionalProposal {
...
 
     pub fn finalize_proposal(
         mut self,
         wallet_process_psbt: impl Fn(&Psbt) -> Result<Psbt, Error>,
         min_feerate_sat_per_vb: Option<FeeRate>,
     ) -> Result<PayjoinProposal, Error> {
+        let sender_input_indexes = self.get_sender_input_indexes();
+        for i in sender_input_indexes {
+            log::trace!("Clearing sender script witness for input {}", i);
+            self.payjoin_psbt.inputs[i].final_script_witness = None;
```
Comment (DanGould) [inline payjoin/src/receive/mod.rs:591]: should final_script_sig and tap_key_sig also be removed at this step? Might they cause problems if they were not removed before wallet_process_psbt of various implementations?

File: payjoin/src/receive/mod.rs [hunk 2]
```diff
@@ -543,6 +543,24 @@ impl ProvisionalProposal {
...
+        }
+
+        Ok(PayjoinProposal {
+            payjoin_psbt: self.payjoin_psbt,
+            owned_vouts: self.owned_vouts,
+            params: self.params,
+        })
+    }
+
+    fn get_sender_input_indexes(&self) -> Vec<usize> {
```
Comment (DanGould) [inline payjoin/src/receive/mod.rs:563]: The `get_` prefix is typically not used by rust functions [by convention](https://rust-lang.github.io/api-guidelines/naming.html#getter-names-follow-rust-convention-c-getter). Especially where it's not a getter (returning a field from a struct) in the typical sense here.

File: payjoin/src/receive/mod.rs [hunk 3]
```diff
@@ -559,27 +577,19 @@ impl ProvisionalProposal {
...
+        return sender_input_indexes;
     }
 
     pub fn finalize_proposal(
         mut self,
         wallet_process_psbt: impl Fn(&Psbt) -> Result<Psbt, Error>,
         min_feerate_sat_per_vb: Option<FeeRate>,
     ) -> Result<PayjoinProposal, Error> {
+        let sender_input_indexes = self.get_sender_input_indexes();
+        for i in sender_input_indexes {
```
Comment (DanGould) [inline payjoin/src/receive/mod.rs:589]: nit 
```suggestion
        for i in self.get_sender_input_indexes() {
```

File: payjoin/src/receive/mod.rs [hunk 4]
```diff
@@ -559,27 +577,19 @@ impl ProvisionalProposal {
...
-            self.payjoin_psbt.inputs[i].final_script_sig = None;
-            self.payjoin_psbt.inputs[i].final_script_witness = None;
-            self.payjoin_psbt.inputs[i].tap_key_sig = None;
-        }
-        Ok(PayjoinProposal {
-            payjoin_psbt: self.payjoin_psbt,
-            owned_vouts: self.owned_vouts,
-            params: self.params,
-        })
+        return sender_input_indexes;
```
Comment (DanGould) [inline payjoin/src/receive/mod.rs:580]:
```suggestion
       sender_input_indexes;
```

Usually we omit the return keyword where possible in rust, at least in this repo

**Review by DanGould — COMMENTED** [review]
@grizznaut May you share how you were able to set up and test JoinMarket with this change?

Overall looks like a correct implementation. I ask that we consider in this PR if any other signatures (or other field?) should also be removed before `apply_fee` or `wallet_process_psbt` are called. Or, if the signature contents removal should be done after apply_fee.

**Comment by grizznaut** [comment]
> May you share how you were able to set up and test JoinMarket with this change?

I reused the same setup I briefly described here https://github.com/payjoin/rust-payjoin/issues/51#issuecomment-2065479887, but pulled and built this branch for the payjoin-cli receiver. I can make a more detailed write-up of how I obtained the certificate, setup the nginx proxy, etc. if that would be helpful!

File: payjoin/src/receive/mod.rs (hunk 1 above)
Comment (grizznaut) [reply payjoin/src/receive/mod.rs:590]: It seems reasonable to remove all signatures at this step as they would be invalid anyways. Updated and smoke-tested with another test receive.

**Pull request merged**